web/source/osk/visualKeyboard.ts

```diff
@@ -1680,19 +1694,38 @@ namespace com.keyman.osk {
    *
    * @param   {Object}  key   base key object
    */
-  touchHold(key: KeyElement) {
+  touchHold(key: KeyElement, force?: boolean) {
     // Clear and restart the popup timer
     if(this.subkeyDelayTimer) {
       window.clearTimeout(this.subkeyDelayTimer);
+      // Cancel the potential subkey event.
       this.subkeyDelayTimer = null;
     }
 
     if(typeof key['subKeys'] != 'undefined' && key['subKeys'] != null) {
-      this.subkeyDelayTimer = window.setTimeout(
-        function(this: VisualKeyboard) {
-          this.clearPopup();
-          this.showSubKeys(key);
-        }.bind(this), this.popupDelay);
+      let _this = this;
+
+      // This Promise receives the user's selected subkey, should it exist.
+      // Note:  the OSK's use of this Promise will allow for passive cancellation.
+      //        If cancelled (so, `null`), the Promise ought remain unresolved.
+      let promise = new Promise<text.KeyEvent>(function(resolve, reject) {
+        let timerId = window.setTimeout(
+          function() {
+            // It's no longer deferred; it's being fulfilled.
+            // Even if the actual subkey itself is still async.
+            _this.subkeyDelayTimer = null;
+            _this.clearPopup();
+            _this.showSubKeys(key, resolve);
+          }, force ? 0 : _this.popupDelay);
+
+          _this.subkeyDelayTimer = timerId;
+      }).then(function(keyEvent: text.KeyEvent) {
+        // Allow active cancellation, even if the source should allow passive.
+        // It's an easy and cheap null guard.
+        if(keyEvent) {
+          PreProcessor.raiseKeyEvent(keyEvent);
+        }
```

Comment:
Same as before; I don't understand this comment. `It's an easy and cheap null guard` is not really very helpful in terms of understanding the why.